### PR TITLE
refactor: add friends using react query 

### LINKED ISSUE #224: Refactoring: Use React Query For Fetching Data
### Objective
We want to improve the app's performance and reduce the number of times the same data needs to be fetched.

### Description
- We should convert the hooks like `useFriends` and `useSignal` to fetch data using react query. (this would allow us to cache the query and reset it easily if desired)
-  `fetchMySignal` should be converted to a useMemo called `mySignal` which will combine the data returned by the react query and perform the rest of the computation that it was already doing
- We should also refactor the `StatusContext` to use the new state of `mySignal` coming from the `useSignal`

### Acceptance Criteria
- All the data fetching for friends, status, .. Should be done via react query
- We should convert the functions which are just fetching to states or useMemos which rely on data returned by `useQuery`

### PR DESCRIPTION
**Pull request by igorntwari**
closes #224

### CODE REVIEW

File: screens/Search/index.tsx [hunk 1]
```diff
@@ -18,75 +18,62 @@ import { User } from "@/types"
...
+      queryClient.setQueryData(["users"], context?.previousUsers)
+      console.warn("Error adding friend:", err)
+    },
+    onSettled: () => {
+      queryClient.invalidateQueries({ queryKey: ["users"] })
+      queryClient.invalidateQueries({ queryKey: ["friends"] })
+    },
+  })
+
+  const filteredUsers = users
```
Comment (Odile078) [inline screens/Search/index.tsx:66]: let's make this a useMemo

File: screens/Search/index.tsx [hunk 2]
```diff
@@ -18,75 +18,62 @@ import { User } from "@/types"
...
-        ),
-      )
-      await fetchAllFriends()
-    } catch (error: any) {
-      console.warn("Error adding friend:", error.response.data.message)
-    } finally {
-      setIsAnyFriendLoading(false)
-      setIsLoadingFriend((prev) => ({ ...prev, [user.id]: false }))
-    }
+  const handleAddFriend = (user: User) => {
```
Comment (Odile078) [inline screens/Search/index.tsx:74]: let's make this a useCallback

File: components/lists/Friends.tsx [hunk 3]
```diff
@@ -1,43 +1,34 @@
...
-  })
+  const { allFriends, isLoading } = useFriends()
   const { friendIds } = temporaryStatus
 
-  const updateFriendsList = useCallback(
-    (friendId: string) => {
-      const newFriends = friendIds?.includes(friendId)
-        ? friendIds.filter((id) => id !== friendId)
-        : [...friendIds, friendId]
+  const updateFriendsList = (friendId: string) => {
```
Comment (Odile078) [inline components/lists/Friends.tsx:14]: why was this changed from a useCallback?

File: hooks/useFriends.ts [hunk 4]
```diff
@@ -1,75 +1,50 @@
...
+    (friend: Friend) =>
+      !availableFriends.some(
+        (availableFriend: Friend) => availableFriend.id === friend.id,
+      ),
+  )
 
-  useEffect(() => {
-    fetchAllFriends()
-  }, [fetchAllFriends])
+  const isLoading = isFriendsLoading || isAvailableFriendsLoading
```
[the remark this answers is not in the document]
Comment (igorntwari) [reply hooks/useFriends.ts:41]: we do not need useMemo() for the isLoading calculation here because the computation is trivial and the value is a primitive, so React handles re-renders efficiently, Adding useMemo() would unnecessarily complicate the code without meaningful performance gains

File: screens/Search/index.tsx (hunk 2 above)
Comment (igorntwari) [reply screens/Search/index.tsx:74]: We do not need to use useCallback for the handleAddFriend function in this specific context
- handleAddFriend is defined and used entirely within the FindFriends component. It is not being passed down as a prop to child components
- The handleAddFriend function is simple and doesn't involve expensive calculations or memory-heavy operations
- Any changes made by addFriendMutation are handled by React Query's internal cache, making additional memoization redundant. The isPending state is tied to the mutation, ensuring consistent behavior without requiring useCallback
- Using useCallback unnecessarily can lead to bloated and harder-to-read code without measurable performance gains

File: components/BottomDrawer.tsx [hunk 5]
```diff
@@ -38,7 +38,10 @@ const BottomDrawer = forwardRef<BottomDrawerRef, DrawerProps>((props, ref) => {
 
   const { refetch } = useQuery({
     queryKey: ["fetch-signaling-friends"],
-    queryFn: () => fetchFriends(),
+    queryFn: async () => {
+      const result = await fetchFriends()
+      return result || []
+    },
```
Comment (serapieTuyishime) [inline components/BottomDrawer.tsx:44]: This should be left as it is, and then add placeholder data, or initial data

**Review by Odile078 — APPROVED** [review]

**Review by serapieTuyishime — APPROVED** [review]
